extensions/vault/runtime/src/main/java/io/quarkus/vault/VaultServiceImpl.java

```diff
@@ -0,0 +1,68 @@
+package io.quarkus.vault;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.vault.service.dto.VaultDynamicCredentials;
+import io.quarkus.vault.service.dto.VaultToken;
+
+public class VaultServiceImpl implements VaultService {
+
+    private static final Logger log = Logger.getLogger(VaultServiceImpl.class);
+
+    private VaultBaseProxy vaultProxy;
+    private VaultServerRuntimeConfig serverConfig;
+    private AtomicReference<VaultToken> auth = new AtomicReference<>(null);
+    private ConcurrentHashMap<String, VaultDynamicCredentials> creds = new ConcurrentHashMap<>();
+
+    public VaultServiceImpl(VaultBaseProxy vaultProxy, VaultServerRuntimeConfig serverConfig) {
+        this.vaultProxy = vaultProxy;
+        this.serverConfig = serverConfig;
+    }
+
+    public VaultServiceImpl(VaultBaseProxy vaultProxy, VaultServerRuntimeConfig serverConfig,
+            AtomicReference<VaultToken> auth) {
+        this(vaultProxy, serverConfig);
+        this.auth = auth;
+    }
+
+    @Override
+    public String getToken() {
+        return serverConfig.token.isPresent() ? serverConfig.token.get() : login().client_token;
+    }
+
+    private VaultToken login() {
```

Comment:
Interesting, so here the login is transparent, done as part of the token acquisition. Can the same be done for `VaultProxy` (where the login operations are explicit ?) (assuming `VaultProxy` and `VaultService` can't really become a single class).

Reply:
see comment above.
VaultProxy = low level
VaultService = complexity hidden